### PR TITLE
Add test for Edge(<vector>) with 0/1 bits

### PR DESCRIPTION
**Pull request by cmarqu**
Questa needs an extra `ReadOnly()` after `Edge()`.
Checking behavior for metavalues is for another PR.

See also #2327.

### CODE REVIEW

**Comment by ktbarrett** [comment]
I'll test on Questa tomorrow, then approve.

**Comment by cmarqu** [comment]
> I assume you already tested Cadence simulators?

Yes, I tested Icarus (VPI), Xcelium (VPI. VHPI), Questa (VPI, FLI), VCS (VPI).
In addition, ghdl in CI also passes.
The result for Questa with VHPI would be interesting.

**Comment by ktbarrett** [comment]
Questa with VHPI doesn't work at all right. They haven't implemented what we use for `ReadWrite`, so regular cocotb writes never happen.

File: tests/test_cases/test_cocotb/test_edge_triggers.py [hunk 1]
```diff
@@ -233,3 +233,39 @@ async def wait_falling_edge():
...
+    cocotb.fork(wait_edge())
+
+    dut.stream_in_data <= 0
+    await ClockCycles(dut.clk, 1)
+
+    for val in range(1, 2**len(dut.stream_in_data)-1):
+        # produce an edge by setting a value != 0:
+        dut.stream_in_data <= val
+        await ClockCycles(dut.clk, 1)
+        edge_cnt_exp = edge_cnt_exp + 1
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_edge_triggers.py:262]: Is there some reason you can't precompute what the expected count will be? I figured it would be `2 * (2**len(dut.stream_in_data))`.

File: tests/test_cases/test_cocotb/test_edge_triggers.py [hunk 2]
```diff
@@ -233,3 +233,39 @@ async def wait_falling_edge():
...
+
+    cocotb.fork(wait_edge())
+
+    dut.stream_in_data <= 0
+    await ClockCycles(dut.clk, 1)
+
+    for val in range(1, 2**len(dut.stream_in_data)-1):
+        # produce an edge by setting a value != 0:
+        dut.stream_in_data <= val
+        await ClockCycles(dut.clk, 1)
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_edge_triggers.py:261]: Is there a reason to use `ClockCycles(clk, 1)` over `RisingEdge(clk)`?

File: tests/test_cases/test_cocotb/test_edge_triggers.py [hunk 3]
```diff
@@ -233,3 +233,39 @@ async def wait_falling_edge():
...
+        edge_cnt_exp = edge_cnt_exp + 1
+        # set back to all-0:
+        dut.stream_in_data <= 0
+        await ClockCycles(dut.clk, 1)
+        edge_cnt_exp = edge_cnt_exp + 1
+
+    if cocotb.SIM_NAME.lower().startswith(("modelsim")):
+        assert edge_cnt == 1
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_edge_triggers.py:269]: Big yikes. Why is this? Maybe a comment would be good. I think it would be better to mark this as an expected fail rather than baking it into the test. It will be easier to locate issues with simulators in the future.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 2 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:261]: This is single-step driving the clock, not reacting to it.
A small side-effect is that I can test Xcelium which otherwise would fail because of #1076.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 3 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:269]: Why... no idea. At least it's consistent between VPI and FLI.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 2 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:261]: (Previous comment was BS...)
No, no reason, will change.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 1 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:262]: It's `2 * ((2**len(dut.stream_in_data)-1)-1)` (508) which takes a while to reason about.

Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:262]: It's more about verifying the code is meets our expectations. If there is poor simulator behavior in both loops, the test may pass when it shouldn't (probably unlikely though).

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 3 above)
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_edge_triggers.py:269]: I think this is worth opening an issue about. Is it hanging on the first edge seen? What happens if you put a `ReadOnly` after every watch?

Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py]: Cool, the extra `ReadOnly` fixed it.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 1 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_edge_triggers.py]: Changed.

File: tests/test_cases/test_cocotb/test_edge_triggers.py (hunk 3 above)
Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_edge_triggers.py]: So `Edge(signal)` immediately after `Edge(signal)` doesn't work on Modelsim.

Comment (ktbarrett) [reply tests/test_cases/test_cocotb/test_edge_triggers.py]: That makes no sense since there is a `RisingEdge` after `RisingEdge` in the main process...

**Review by ktbarrett — APPROVED** [review]
